File: backend/services/sms_gateway/gsm_driver.py
"""GSM HAT driver for SMS communication."""
import asyncio
import serial
import re
import logging
from typing import Optional, List, Dict, Any
from datetime import datetime
import gammu
from backend.shared.config import settings

logger = logging.getLogger(__name__)


class GSMDriver:
    """GSM HAT driver using Gammu library."""
    
    def __init__(self):
        self.state_machine = None
        self.is_connected = False
        self.signal_strength = 0
        self.network_name = ""
        self.phone_number = ""
        
    async def initialize(self) -> bool:
        """Initialize GSM connection."""
        try:
            # Configure Gammu state machine
            self.state_machine = gammu.StateMachine()
            
            # Try different device configurations
            configs = [
                {
                    'Device': settings.sms_device,
                    'Connection': 'serial',
                    'BaudRate': settings.sms_baud_rate,
                    'DataBits': 8,
                    'StopBits': 1,
                    'Parity': 'None',
                    'FlowControl': 'None'
                },
                {
                    'Device': '/dev/ttyACM0',  # Alternative device path
                    'Connection': 'serial',
                    'BaudRate': 115200,
                    'DataBits': 8,
                    'StopBits': 1,
                    'Parity': 'None',
                    'FlowControl': 'None'
                }
            ]
            
            for config in configs:
                try:
                    self.state_machine.ReadConfig(config)
                    await asyncio.sleep(1)  # Wait for connection
                    
                    # Test connection
                    self.state_machine.Init()
                    self.is_connected = True
                    
                    # Get device info
                    await self._get_device_info()
                    
                    logger.info(f"GSM connection established on {config['Device']}")
                    return True
                    
                except Exception as e:
                    logger.warning(f"Failed to connect on {config['Device']}: {e}")
                    continue
            
            logger.error("Failed to establish GSM connection on any device")
            return False
            
        except Exception as e:
            logger.error(f"GSM initialization failed: {e}")
            return False
    
    async def _get_device_info(self) -> None:
        """Get device information."""
        try:
            # Get network info
            network_info = self.state_machine.GetNetworkInfo()
            self.network_name = network_info.get('NetworkName', 'Unknown')
            
            # Get signal strength
            signal_info = self.state_machine.GetSignalQuality()
            self.signal_strength = signal_info.get('SignalPercent', 0)
            
            # Get phone number
            identity = self.state_machine.GetIMEI()
            self.phone_number = identity.get('IMEI', 'Unknown')
            
            logger.info(f"Network: {self.network_name}, Signal: {self.signal_strength}%, IMEI: {self.phone_number}")
            
        except Exception as e:
            logger.warning(f"Could not get device info: {e}")
    
    async def send_sms(self, phone_number: str, message: str) -> bool:
        """Send SMS message."""
        if not self.is_connected:
            logger.error("GSM not connected")
            return False
        
        try:
            # Validate phone number format
            phone_number = self._format_phone_number(phone_number)
            if not phone_number:
                logger.error(f"Invalid phone number format: {phone_number}")
                return False
            
            # Check message length
            if len(message) > 160:
                logger.error(f"Message too long: {len(message)} characters")
                return False
            
            # Create SMS message
            sms_message = {
                'Text': message,
                'SMSC': {'Location': 1},
                'Number': phone_number,
            }
            
            # Send SMS
            self.state_machine.SendSMS(sms_message)
            
            logger.info(f"SMS sent to {phone_number}: {message[:50]}...")
            return True
            
        except Exception as e:
            logger.error(f"Failed to send SMS: {e}")
            return False
    
    async def receive_sms(self) -> List[Dict[str, Any]]:
        """Receive pending SMS messages."""
        if not self.is_connected:
            logger.error("GSM not connected")
            return []
        
        try:
            messages = []
            
            # Get SMS memory status
            status = self.state_machine.GetSMSFoldersStatus()
            
            for folder in status:
                if folder['New'] > 0:
                    # Read messages from this folder
                    start = True
                    
                    while True:
                        try:
                            if start:
                                sms_list = self.state_machine.GetNextSMS(Start=True, Folder=folder['Folder'])
                                start = False
                            else:
                                sms_list = self.state_machine.GetNextSMS(Location=sms_list[0]['Location'], Folder=folder['Folder'])
                            
                            for sms in sms_list:
                                message = {
                                    'id': f"sms_{sms['Location']}_{datetime.utcnow().timestamp()}",
                                    'sender': sms['Number'],
                                    'content': sms['Text'],
                                    'timestamp': sms['DateTime'],
                                    'folder': folder['Folder'],
                                    'location': sms['Location']
                                }
                                messages.append(message)
                                
                        except gammu.ERR_EMPTY:
                            break
                        except Exception as e:
                            logger.error(f"Error reading SMS: {e}")
                            break
            
            if messages:
                logger.info(f"Received {len(messages)} new SMS messages")
            
            return messages
            
        except Exception as e:
            logger.error(f"Failed to receive SMS: {e}")
            return []
    
    async def delete_sms(self, folder: int, location: int) -> bool:
        """Delete SMS message."""
        if not self.is_connected:
            return False
        
        try:
            self.state_machine.DeleteSMS(Folder=folder, Location=location)
            logger.info(f"Deleted SMS from folder {folder}, location {location}")
            return True
        except Exception as e:
            logger.error(f"Failed to delete SMS: {e}")
            return False
    
    async def get_signal_strength(self) -> int:
        """Get current signal strength."""
        if not self.is_connected:
            return 0
        
        try:
            signal_info = self.state_machine.GetSignalQuality()
            self.signal_strength = signal_info.get('SignalPercent', 0)
            return self.signal_strength
        except Exception as e:
            logger.error(f"Failed to get signal strength: {e}")
            return 0
    
    async def is_network_available(self) -> bool:
        """Check if network is available."""
        if not self.is_connected:
            return False
        
        try:
            network_info = self.state_machine.GetNetworkInfo()
            return network_info.get('State') == 'Registered'
        except Exception as e:
            logger.error(f"Failed to check network status: {e}")
            return False
    
    def _format_phone_number(self, phone_number: str) -> Optional[str]:
        """Format phone number for SMS sending."""
        # Remove all non-digit characters
        digits = re.sub(r'\D', '', phone_number)
        
        # Add country code if missing
        if len(digits) == 10:
            digits = '+1' + digits  # Assume US number
        elif len(digits) == 11 and digits.startswith('1'):
            digits = '+' + digits
        elif not digits.startswith('+'):
            digits = '+' + digits
        
        # Validate format
        if len(digits) < 10:
            return None
        
        return digits
    
    async def disconnect(self) -> None:
        """Disconnect GSM connection."""
        if self.is_connected and self.state_machine:
            try:
                self.state_machine.Terminate()
                self.is_connected = False
                logger.info("GSM connection terminated")
            except Exception as e:
                logger.error(f"Error disconnecting GSM: {e}")
    
    async def health_check(self) -> Dict[str, Any]:
        """Get GSM health status."""
        return {
            'connected': self.is_connected,
            'signal_strength': self.signal_strength,
            'network_name': self.network_name,
            'phone_number': self.phone_number,
            'network_available': await self.is_network_available() if self.is_connected else False
        }


class SerialGSMDriver:
    """Alternative GSM driver using direct serial communication."""
    
    def __init__(self):
        self.serial_connection = None
        self.is_connected = False
        
    async def initialize(self) -> bool:
        """Initialize serial connection to GSM module."""
        try:
            # Try different serial configurations
            configs = [
                {'port': settings.sms_device, 'baudrate': settings.sms_baud_rate},
                {'port': '/dev/ttyACM0', 'baudrate': 115200},
                {'port': '/dev/ttyUSB1', 'baudrate': 115200},
            ]
            
            for config in configs:
                try:
                    self.serial_connection = serial.Serial(
                        port=config['port'],
                        baudrate=config['baudrate'],
                        timeout=1,
                        write_timeout=1
                    )
                    
                    # Test connection with AT command
                    await self._send_at_command('AT')
                    await asyncio.sleep(1)
                    
                    # Check if GSM module responds
                    response = await self._send_at_command('AT+CPIN?')
                    if 'READY' in response:
                        self.is_connected = True
                        logger.info(f"Serial GSM connection established on {config['port']}")
                        return True
                    
                except Exception as e:
                    logger.warning(f"Failed to connect on {config['port']}: {e}")
                    if self.serial_connection:
                        self.serial_connection.close()
                    continue
            
            logger.error("Failed to establish serial GSM connection")
            return False
            
        except Exception as e:
            logger.error(f"Serial GSM initialization failed: {e}")
            return False
    
    async def _send_at_command(self, command: str, timeout: float = 5.0) -> str:
        """Send AT command and get response."""
        if not self.serial_connection:
            raise Exception("Serial connection not established")
        
        try:
            # Clear input buffer
            self.serial_connection.reset_input_buffer()
            
            # Send command
            self.serial_connection.write(f"{command}\r\n".encode())
            
            # Read response
            response = ""
            start_time = asyncio.get_event_loop().time()
            
            while (asyncio.get_event_loop().time() - start_time) < timeout:
                if self.serial_connection.in_waiting > 0:
                    data = self.serial_connection.read(self.serial_connection.in_waiting)
                    response += data.decode('utf-8', errors='ignore')
                    
                    # Check for end of response
                    if 'OK' in response or 'ERROR' in response:
                        break
                
                await asyncio.sleep(0.1)
            
            return response.strip()
            
        except Exception as e:
            logger.error(f"AT command failed: {e}")
            return ""
    
    async def send_sms(self, phone_number: str, message: str) -> bool:
        """Send SMS using AT commands."""
        if not self.is_connected:
            return False
        
        try:
            # Set SMS text mode
            await self._send_at_command('AT+CMGF=1')
            
            # Set recipient number
            await self._send_at_command(f'AT+CMGS="{phone_number}"')
            
            # Send message content
            self.serial_connection.write(f"{message}\x1A".encode())
            await asyncio.sleep(2)
            
            # Check response
            response = ""
            start_time = asyncio.get_event_loop().time()
            while (asyncio.get_event_loop().time() - start_time) < 10:
                if self.serial_connection.in_waiting > 0:
                    data = self.serial_connection.read(self.serial_connection.in_waiting)
                    response += data.decode('utf-8', errors='ignore')
                    
                    if 'OK' in response:
                        logger.info(f"SMS sent to {phone_number}")
                        return True
                    elif 'ERROR' in response:
                        logger.error(f"SMS send failed: {response}")
                        return False
                
                await asyncio.sleep(0.1)
            
            logger.error("SMS send timeout")
            return False
            
        except Exception as e:
            logger.error(f"Failed to send SMS: {e}")
            return False
    
    async def disconnect(self) -> None:
        """Disconnect serial connection."""
        if self.serial_connection and self.is_connected:
            try:
                self.serial_connection.close()
                self.is_connected = False
                logger.info("Serial GSM connection closed")
            except Exception as e:
                logger.error(f"Error closing serial connection: {e}")

File: backend/rust_services/sms_gateway/src/gsm_driver.rs
//! GSM HAT driver for SIM800C/SIM7000 modules
//! 
//! Provides low-level AT command interface for GSM communication

use crate::error::{GatewayError, Result};
use async_trait::async_trait;
use chrono::{DateTime, Utc};
use std::time::{Duration, Instant};
use tokio::time::timeout;
use tokio::io::{AsyncReadExt, AsyncWriteExt};
use tokio_serial::{SerialPort, SerialPortBuilderExt, SerialStream};
use tracing::{debug, info, warn};

/// GSM driver trait for different implementations
#[async_trait]
pub trait GSMDriver: Send + Sync {
    /// Initialize the GSM connection
    async fn initialize(&mut self) -> Result<()>;
    
    /// Check if connected
    fn is_connected(&self) -> bool;
    
    /// Send SMS message
    async fn send_sms(&mut self, phone_number: &str, message: &str) -> Result<()>;
    
    /// Receive pending SMS messages
    async fn receive_sms(&mut self) -> Result<Vec<ReceivedSMS>>;
    
    /// Get signal strength (0-100)
    async fn get_signal_strength(&mut self) -> Result<u8>;
    
    /// Get network name
    async fn get_network_name(&mut self) -> Result<String>;
    
    /// Check if network is registered
    async fn is_network_registered(&mut self) -> Result<bool>;
    
    /// Disconnect from GSM module
    async fn disconnect(&mut self) -> Result<()>;
}

/// Received SMS message structure
#[derive(Debug, Clone)]
pub struct ReceivedSMS {
    pub id: String,
    pub sender: String,
    pub content: String,
    pub timestamp: DateTime<Utc>,
    pub folder: u8,
    pub location: u16,
}

/// Serial-based GSM driver using AT commands
pub struct SerialGSMDriver {
    port: Option<SerialStream>,
    config: SerialDriverConfig,
    last_command_time: Option<Instant>,
}

struct SerialDriverConfig {
    device: String,
    baud_rate: u32,
    timeout: Duration,
    min_command_interval: Duration,
}

impl SerialDriverConfig {
    fn new(device: String, baud_rate: u32) -> Self {
        Self {
            device,
            baud_rate,
            timeout: Duration::from_secs(5),
            min_command_interval: Duration::from_millis(100),
        }
    }
}

impl SerialGSMDriver {
    /// Create a new serial GSM driver
    pub fn new(device: String, baud_rate: u32) -> Self {
        Self {
            port: None,
            config: SerialDriverConfig::new(device, baud_rate),
            last_command_time: None,
        }
    }
    
    /// Send AT command and wait for response
    async fn send_at_command(&mut self, command: &str) -> Result<String> {
        // Rate limiting: ensure minimum interval between commands
        if let Some(last_time) = self.last_command_time {
            let elapsed = last_time.elapsed();
            if elapsed < self.config.min_command_interval {
                tokio::time::sleep(self.config.min_command_interval - elapsed).await;
            }
        }
        
        let port = self.port.as_mut().ok_or(GatewayError::NotConnected)?;
        
        // Clear input buffer
        let _ = port.clear(tokio_serial::ClearBuffer::Input);
        
        // Send command
        let cmd_bytes = format!("{}\r\n", command);
        port.write_all(cmd_bytes.as_bytes())
            .await
            .map_err(|e| GatewayError::Io(e))?;
        
        self.last_command_time = Some(Instant::now());
        
        // Read response with timeout
        let mut response = String::new();
        let mut buffer = [0u8; 256];
        let start = Instant::now();
        
        while start.elapsed() < self.config.timeout {
            match timeout(Duration::from_millis(100), port.read(&mut buffer)).await {
                Ok(Ok(n)) if n > 0 => {
                    let text = String::from_utf8_lossy(&buffer[..n]);
                    response.push_str(&text);
                    
                    // Check for end of response
                    if response.contains("OK") || response.contains("ERROR") {
                        break;
                    }
                }
                Ok(Ok(_)) => {
                    tokio::time::sleep(Duration::from_millis(10)).await;
                }
                Ok(Err(e)) => {
                    return Err(GatewayError::Io(e));
                }
                Err(_) => {
                    // Timeout on read, continue waiting
                    tokio::time::sleep(Duration::from_millis(10)).await;
                }
            }
        }
        
        if response.is_empty() {
            return Err(GatewayError::Timeout(format!(
                "No response to command: {}",
                command
            )));
        }
        
        debug!("AT command: {} -> {}", command, response.trim());
        Ok(response)
    }
    
    /// Wait for specific response
    async fn wait_for_response(&mut self, expected: &str) -> Result<()> {
        let mut buffer = [0u8; 256];
        let start = Instant::now();
        
        while start.elapsed() < self.config.timeout {
            match timeout(Duration::from_millis(100), self.port.as_mut().unwrap().read(&mut buffer)).await {
                Ok(Ok(n)) if n > 0 => {
                    let text = String::from_utf8_lossy(&buffer[..n]);
                    if text.contains(expected) {
                        return Ok(());
                    }
                    if text.contains("ERROR") {
                        return Err(GatewayError::AtCommandFailed(text.to_string()));
                    }
                }
                _ => {
                    tokio::time::sleep(Duration::from_millis(10)).await;
                }
            }
        }
        
        Err(GatewayError::Timeout(format!("Expected: {}", expected)))
    }
    
    /// Format phone number for SMS
    fn format_phone_number(&self, phone: &str) -> Result<String> {
        // Remove all non-digit characters
        let digits: String = phone.chars().filter(|c| c.is_ascii_digit()).collect();
        
        if digits.len() < 10 {
            return Err(GatewayError::InvalidPhoneNumber(phone.to_string()));
        }
        
        // Add country code if needed
        let formatted = if digits.len() == 10 {
            format!("+1{}", digits)
        } else if digits.len() == 11 && digits.starts_with('1') {
            format!("+{}", digits)
        } else if !digits.starts_with('+') {
            format!("+{}", digits)
        } else {
            format!("+{}", digits)
        };
        
        Ok(formatted)
    }
}

#[async_trait]
impl GSMDriver for SerialGSMDriver {
    async fn initialize(&mut self) -> Result<()> {
        info!("Initializing serial GSM driver on {}", self.config.device);
        
        // Try to open serial port
        let port = tokio_serial::new(&self.config.device, self.config.baud_rate)
            .open_native_async()
            .map_err(|e| GatewayError::Serial(e))?;
        
        self.port = Some(port);
        
        // Wait for module to be ready
        tokio::time::sleep(Duration::from_secs(2)).await;
        
        // Test connection with AT command
        let response = self.send_at_command("AT").await?;
        if !response.contains("OK") {
            return Err(GatewayError::InitializationFailed(
                "AT command failed".to_string(),
            ));
        }
        
        // Check if SIM card is ready
        let response = self.send_at_command("AT+CPIN?").await?;
        if !response.contains("READY") {
            warn!("SIM card may not be ready: {}", response);
        }
        
        // Set SMS text mode
        let response = self.send_at_command("AT+CMGF=1").await?;
        if !response.contains("OK") {
            return Err(GatewayError::InitializationFailed(
                "Failed to set SMS text mode".to_string(),
            ));
        }
        
        // Get network info
        let _ = self.get_network_name().await;
        let _ = self.get_signal_strength().await;
        
        info!("Serial GSM driver initialized successfully");
        Ok(())
    }
    
    fn is_connected(&self) -> bool {
        self.port.is_some()
    }
    
    async fn send_sms(&mut self, phone_number: &str, message: &str) -> Result<()> {
        if !self.is_connected() {
            return Err(GatewayError::NotConnected);
        }
        
        // Validate message length
        if message.len() > 160 {
            return Err(GatewayError::MessageTooLong(message.len()));
        }
        
        // Format phone number
        let formatted_number = self.format_phone_number(phone_number)?;
        
        info!("Sending SMS to {}: {}...", formatted_number, &message[..message.len().min(50)]);
        
        // Set SMS text mode (ensure it's set)
        self.send_at_command("AT+CMGF=1").await?;
        
        // Set recipient number
        let cmd = format!("AT+CMGS=\"{}\"", formatted_number);
        let response = self.send_at_command(&cmd).await?;
        
        if !response.contains(">") {
            return Err(GatewayError::AtCommandFailed(
                "Did not receive prompt for message".to_string(),
            ));
        }
        
        // Send message content with Ctrl+Z terminator
        let port = self.port.as_mut().unwrap();
        let message_bytes = format!("{}\x1A", message);
        port.write_all(message_bytes.as_bytes())
            .await
            .map_err(|e| GatewayError::Io(e))?;
        
        // Wait for confirmation
        self.wait_for_response("OK").await?;
        
        info!("SMS sent successfully to {}", formatted_number);
        Ok(())
    }
    
    async fn receive_sms(&mut self) -> Result<Vec<ReceivedSMS>> {
        if !self.is_connected() {
            return Err(GatewayError::NotConnected);
        }
        
        // List all unread SMS
        let response = self.send_at_command("AT+CMGL=\"REC UNREAD\"").await?;
        
        if response.contains("ERROR") {
            return Err(GatewayError::AtCommandFailed(response));
        }
        
        // Parse SMS messages from response
        let mut messages = Vec::new();
        let lines: Vec<&str> = response.lines().collect();
        
        let mut i = 0;
        while i < lines.len() {
            // Look for SMS header line (e.g., "+CMGL: 0,\"REC UNREAD\",\"+1234567890\",\"\",\"23/12/15,10:30:00+00\"")
            if lines[i].starts_with("+CMGL:") {
                // Parse header
                let parts: Vec<&str> = lines[i].split(',').collect();
                if parts.len() >= 5 {
                    let location = parts[0].split(':').nth(1).and_then(|s| s.trim().parse().ok()).unwrap_or(0);
                    let sender = parts[2].trim_matches('"');
                    
                    // Next line should be the message content
                    if i + 1 < lines.len() {
                        let content = lines[i + 1].trim().to_string();
                        let timestamp = Utc::now(); // TODO: Parse actual timestamp from header
                        
                        messages.push(ReceivedSMS {
                            id: format!("sms_{}_{}", location, timestamp.timestamp()),
                            sender: sender.to_string(),
                            content,
                            timestamp,
                            folder: 1, // Inbox
                            location: location as u16,
                        });
                        
                        i += 2;
                        continue;
                    }
                }
            }
            i += 1;
        }
        
        if !messages.is_empty() {
            info!("Received {} SMS messages", messages.len());
        }
        
        Ok(messages)
    }
    
    async fn get_signal_strength(&mut self) -> Result<u8> {
        if !self.is_connected() {
            return Err(GatewayError::NotConnected);
        }
        
        let response = self.send_at_command("AT+CSQ").await?;
        
        // Parse response: +CSQ: <rssi>,<ber>
        // RSSI: 0-31 (99 = not known/not detectable)
        // Convert to percentage: (rssi / 31) * 100
        if let Some(start) = response.find("+CSQ:") {
            let rest = &response[start + 5..];
            if let Some(comma) = rest.find(',') {
                if let Ok(rssi) = rest[..comma].trim().parse::<u8>() {
                    if rssi == 99 {
                        return Ok(0);
                    }
                    let percentage = (rssi as f32 / 31.0 * 100.0) as u8;
                    return Ok(percentage.min(100));
                }
            }
        }
        
        Ok(0)
    }
    
    async fn get_network_name(&mut self) -> Result<String> {
        if !self.is_connected() {
            return Err(GatewayError::NotConnected);
        }
        
        let response = self.send_at_command("AT+COPS?").await?;
        
        // Parse response: +COPS: <mode>[,<format>,<oper>]
        if let Some(start) = response.find("+COPS:") {
            let rest = &response[start + 6..];
            let parts: Vec<&str> = rest.split(',').collect();
            if parts.len() >= 3 {
                let name = parts[2].trim_matches('"');
                return Ok(name.to_string());
            }
        }
        
        Ok("Unknown".to_string())
    }
    
    async fn is_network_registered(&mut self) -> Result<bool> {
        if !self.is_connected() {
            return Err(GatewayError::NotConnected);
        }
        
        let response = self.send_at_command("AT+CREG?").await?;
        
        // Parse response: +CREG: <n>,<stat>
        // stat: 0=not registered, 1=registered (home), 2=searching, 3=denied, 5=registered (roaming)
        if let Some(start) = response.find("+CREG:") {
            let rest = &response[start + 6..];
            if let Some(comma) = rest.find(',') {
                if let Ok(stat) = rest[comma + 1..].trim().parse::<u8>() {
                    return Ok(stat == 1 || stat == 5);
                }
            }
        }
        
        Ok(false)
    }
    
    async fn disconnect(&mut self) -> Result<()> {
        if let Some(mut port) = self.port.take() {
            // Send AT command to put module in low power mode (optional)
            let _ = self.send_at_command("AT+CPOWD=1").await;
            
            // Close port
            port.shutdown().await.map_err(|e| GatewayError::Io(e))?;
            info!("GSM driver disconnected");
        }
        Ok(())
    }
}

